### PR TITLE
fix Ink not updating on TextField newline

### LINKED ISSUE #138219: ListView doesn't rerender ListTile border on TextFormField line added
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Build the app in the code sample
2. Add new line in the text field at the top
3. see the text field border intersecting with the `ListTile`

### Expected results

The border of the `ListTile` should move downwards when adding a new line, like when using a `Column`.

In case not updating is intended, the `ListTile`  title should behave like the border.

### Actual results

The `ListTile` border stays in place:
image

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ListView(
        children: <Widget>[
          TextFormField(
            minLines: 1,
            maxLines: 4,
          ),
          const ListTile(
            title: Text('moving text'),
            shape: RoundedRectangleBorder(
              side: BorderSide(
                color: Colors.red
              )
            ),
          )
        ],
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/82763757/329eaf0d-e683-4916-a72b-9e6010acc720

</details>

_(Local playback works fine)_

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.8, on Garuda Linux 6.6.1-zen1-1-zen, locale en_US.UTF-8)
    • Flutter version 3.13.8 on channel stable at /home/derdilla/bin/apps/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c4930c4ac (3 weeks ago), 2023-10-18 10:57:55 -0500
    • Engine revision 767d8c75e8
    • Dart version 3.1.4
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /home/derdilla/Android/Sdk
    • Platform android-34, build-tools 34.0.0-rc3
    • ANDROID_HOME = /home/derdilla/Android/Sdk
    • Java binary at: /home/derdilla/bin/apps/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 16.0.6
    • cmake version 3.27.7
    • ninja version 1.11.1
    • pkg-config version 1.8.1

[✓] Android Studio (version 2023.1)
    • Android Studio at /home/derdilla/bin/apps/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Android Studio (version 2022.3)
    • Android Studio at /home/derdilla/.local/share/JetBrains/Toolbox/apps/android-studio
    • Flutter plugin version 75.1.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at /home/derdilla/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate
    • Flutter plugin version 75.1.4
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (3 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 14 (API 34) (emulator)
    • Linux (desktop)              • linux         • linux-x64      • Garuda Linux 6.6.1-zen1-1-zen
    • Chrome (web)                 • chrome        • web-javascript • Chromium 119.0.6045.123 Arch Linux

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

Also works on main branch.

### PR DESCRIPTION
**Pull request by derdilla**
## Summary

As documented in the [Ink](https://api.flutter.dev/flutter/material/Ink-class.html#limitations) class, a layout [LayoutChangedNotification](https://api.flutter.dev/flutter/widgets/LayoutChangedNotification-class.html) needs to be dispatched.
When a `FormField` receives inputs it may change its size which causes box decorations (which are based on InkWells) on other elements not to update. To address this issue the text form field now dispatches the notification when it needs to be rebuild which must happen on a size change.

This change is tested by adding an Ink below a text field to which a new line is added.

Fixes #138219

### Old behavior:

material text_form_field update_decorations

### New behavior:

material text_form_field update_decorations

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing. (in text_form_field_test.dart)

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by justinmc — COMMENTED** [review]
Thanks for the PR! This is an interesting bug. I want to make sure SizeChangedLayoutNotifier is the right approach.

@goderbauer Is it an oversight that we've never added SizeChangedLayoutNotifier to EditableText before? It seems that Ink doesn't update its position when the size of an EditableText changes:

<table>
<tr>
<td></td>
<td>
<code>
TextField(minLines: 1, maxLines: 4),
ListTile(
  title: Text('moving text'),
  shape: RoundedRectangleBorder(
    side: BorderSide(
      color: Colors.red
    )
  ),
),
</code>
</td>
</tr>
</table>

Or is there something more fundamental to solve here?

File: packages/flutter/test/material/text_form_field_test.dart
Comment (NobodyForNothing) [reply packages/flutter/test/material/text_form_field_test.dart:1676]: I don't know if testing without a golden is possible. The ink (and children) get moved correctly, but the render object doesn't repaint. 

One test I could think of would be to listen for the dispatch of a SizeChangedLayoutNotification. Do you think that would be sufficient?

Comment (justinmc) [reply packages/flutter/test/material/text_form_field_test.dart:1676]: Yes that sounds better than a golden to me, if indeed getting the position doesn't work here then listening for the notification should be good enough.

Comment (NobodyForNothing) [reply packages/flutter/test/material/text_form_field_test.dart]: @justinmc PR is ready for another review.

**Comment by goderbauer** [comment]
> Is it an oversight that we've never added SizeChangedLayoutNotifier to EditableText before?

I am not entirely sure. My guess is that this was just an oversight and nobody thought about this before.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:16757]: Before this line, can you expect that the count is zero? Just to be clear that the notification happens due to the entered text and not during initial build.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just a nit in the test. Let's get one more reviewer and then we can merge.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (bleroux) [inline packages/flutter/test/widgets/editable_text_test.dart:16755]: Nit, Replace `),);` with `));`

**Review by bleroux — APPROVED** [review]
LGTM!
Thanks for your contribution.

Before we merge this thank you to address the last 2 nits.

**Pull request merged**
